README.md

```diff
@@ -33,106 +33,131 @@ A project was created with those packages `npm install`ed, creating a gigantic `
+These are currently unsupported but reserved for future use. Browsers support specifiers like `'/app.js'` to be relative to the base URL of the page, whereas in CommonJS a specifier starting with `/` refers to the root of the file system. We would like to find a solution that conforms Node closer to browsers for `/`-leading specifiers.
 
-```js
-import _ from 'underscore';
-// underscore has a package.json "main" field specifying "underscore.js"
-
-import pad from 'pad';
-// pad has a file named index.js in its root
-```
+### `createRequireFromPath`
 
-In other words, the specifiers here behave the same as if they were in `require` calls, because the package name (`underscore` or `pad`) is detected as a CommonJS package. Packages are detected as CommonJS by _lacking_ a signifier that the package should be treated as ESM, for example the `"exports"` field from the [package exports proposal](https://github.com/jkrems/proposal-pkg-exports) or another field like [`"mode"`](https://github.com/nodejs/node/pull/18392).
+Packages or files located from `require` calls created via [`module.createRequireFromPath`](https://nodejs.org/docs/latest/api/modules.html#modules_module_createrequirefrompath_filename) are *always* parsed as CommonJS, following how `require` behaves now.
 
-“Dual-mode” packages, that export both ESM and CommonJS entry points, are loaded as ESM when imported via the `import` statement, and follow the rules of the package exports proposal. To explicitly import a dual-mode package via its CommonJS entry point, [`module.createRequireFromPath`](https://nodejs.org/docs/latest/api/modules.html#modules_module_createrequirefrompath_filename) could be used.
+### “Dual mode” packages
 
+A package can be “dual mode” if its `package.json` contains both a `"main"` field and an `"exports"` field (or some other ESM-signifying field). An `import` statement of such a package will treat the package as ESM and ignore the `"main"` field. To explicitly import a dual-mode package via its CommonJS entry point, [`module.createRequireFromPath`](https://nodejs.org/docs/latest/api/modules.html#modules_module_createrequirefrompath_filename) could be used.
 
-#### `import` statements of CommonJS files within CommonJS packages (“deep imports”)
+The ESM and CommonJS versions of a dual-mode package are really two distinct packages, and would be treated as such by Node if both were imported. Unless there is a legitimate use case for such behavior, though, we expect Node would throw an exception if a user tries to import the same package in both modes into the same package scope. Different modes of the same package _should_ be importable if a package boundary separates them, for example if a user’s project imports ESM `lodash` and that project has a dependency which itself imports CommonJS `lodash`.
```

Comment:
Is the expectation that the parent is now affecting access control to modules? This is a new concept...

Reply:
I think we can rationalise the dual identities by saying that package.json forwards to (two, or more) independent files/modules, i.e. package identity doesn't exist, module identity is what matters. 

The corrolary of this is that, for dual-mode packages, we must prevent ESM entrypoints colliding with the main/CJS entrypoint, as that would result in a single identity for the package.

Reply:
> I think we can rationalise the dual identities by saying that package.json forwards to (two, or more) independent files/modules, i.e. package identity doesn’t exist, module identity is what matters.

By “module” here, do you mean a package (a folder with source files and a `package.json`) or a single file, like a `.js` file that exports something? I’m not really following this sentence.

I look at it as a “dual-mode” package really being two packages, based on its two entry points. This is no different really than the lots of packages today that have multiple entry points, like the `"browser"` key that points to a browser build. They’re separate “packages” that happen to overlap in the same folder with the same `package.json`.

> for dual-mode packages, we must prevent ESM entrypoints colliding with the main/CJS entrypoint, as that would result in a single identity for the package.

Say you _did_ create a `package.json` with `"main": "index.js"` and `"exports": "index.js"`, following the package exports proposal. If `index.js` contained any `import` or `export` statements, `require`ing the package from CommonJS would fail; likewise if `index.js` contained any references to the CommonJS globals like `require` or `module.exports`. I suppose it could guard those references, like `if (exports) exports = ...`, so that it wouldn’t throw in ESM mode but would actually export something in CommonJS mode. But really at this point it’s people trying to game the system, and since Node already lets them override the module loading machinery I’m not sure it’s worth it to try to hard to stomp out such things. We could stop it by throwing an error if the ESM and CJS entry points both point to the same file, if we want to be very safe.

Reply:
> Is the expectation that the parent is now affecting access control to modules? This is a new concept…

How so? What does “parent” mean here?

All I meant by this is that say I have an app like this:

```
~/app/package.json
~/app/app.js
~/app/node_modules/lodash (dual-mode ESM/CJS package)
~/app/node_modules/request/package.json
~/app/node_modules/request/index.js
~/app/node_modules/request/node_modules/lodash (dual-mode ESM/CJS package)
```

In `app.js`, my ESM app has `import 'lodash'`. And then in `request/index.js`, Request has `require('lodash')`. In the context of my app, I just pulled in the ESM `lodash`; while Request gets the CJS `lodash`. The two imports are in separate package scopes, so the two packages get different `lodash`es.

Or say it’s current Node and everything is CJS, but the same folders and files as above. `~/app/package.json` includes `"lodash": "2.0.0"` while `request/package.json` includes `"lodash": "1.0.0"`. My app would get Lodash 2.0.0, while Request would get Lodash 1.0.0. We already have this concept of separate dependency versions in separate package scopes; we would just be extending it from version numbers to include parse goals (ESM or CJS).

And just as my app couldn’t `require` both Lodash 1.0.0 and 2.0.0, it shouldn’t be able to `import`/`require` both the ESM and CJS versions of Lodash.

Reply:
> And just as my app couldn’t require both Lodash 1.0.0 and 2.0.0, it shouldn’t be able to import/require both the ESM and CJS versions of Lodash.

Your app can require both Lodash 1.0.0 and Lodash 2.0.0 if they are installed locally under different aliases.

The problem with this one is that its a **race** as to which one will get the valid package and which one will error, which seems brittle, error-prone and hard to debug. I don't know that there is a way to tackle this without it being a race, unless we entirely disable one or the other (which was my comment).

Reply:
Can you explain how a race could happen? We were discussing this in #1. Wouldn't Node resolve all the `import` statements before evaluating any code, and therefore if the code contains any `require`s they'll always be evaluated second?

Reply:
Because you can have dynamic import. You can also run CommonJS first via `-r` or other process hooks like in loaders.

Reply:
Can we shift the discussion of “what should we do in this case” to #1? And keep this focused on what the language in this PR should say. For now, how about I change the language to something like “it remains to be decided what Node should do in this case.”